File: go/examples/stream-write-example.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"sync/atomic"
	"syscall"
	"time"

	laserstream "github.com/helius-labs/laserstream-sdk/go"
	"github.com/joho/godotenv"
)

func main() {
	// Get configuration from environment
	godotenv.Load("../.env")

	endpoint := os.Getenv("LASERSTREAM_PRODUCTION_ENDPOINT")
	if endpoint == "" {
		log.Fatal("LASERSTREAM_PRODUCTION_ENDPOINT required")
	}
	apiKey := os.Getenv("LASERSTREAM_PRODUCTION_API_KEY")
	if apiKey == "" {
		log.Fatal("LASERSTREAM_PRODUCTION_API_KEY required")
	}

	// Create client with compression enabled
	channelOptions := &laserstream.ChannelOptions{
		UseCompression: true,
	}

	config := laserstream.LaserstreamConfig{
		Endpoint:       endpoint,
		APIKey:         apiKey,
		ChannelOptions: channelOptions,
	}

	// Create client
	client := laserstream.NewClient(config)
	commitmentLevel := laserstream.CommitmentLevel_FINALIZED

	// Initial subscription request - just subscribe to slots
	initialReq := &laserstream.SubscribeRequest{
		Slots: map[string]*laserstream.SubscribeRequestFilterSlots{
			"client": {
				FilterByCommitment: &[]bool{true}[0],
			},
		},
		Commitment: &commitmentLevel,
	}

	fmt.Println("🚀 Laserstream Bidirectional Stream Example")
	fmt.Println("📡 Starting initial subscription to slots...")

	var messageCount int32
	var transactionAdded atomic.Bool
	var blockAdded atomic.Bool

	// Data handler
	dataHandler := func(update *laserstream.SubscribeUpdate) {
		count := atomic.AddInt32(&messageCount, 1)

		switch u := update.UpdateOneof.(type) {
		case *laserstream.SubscribeUpdate_Slot:
			fmt.Printf("🎰 Slot update #%d: %d\n", count, u.Slot.Slot)

			// After receiving 5 slot updates, add a transaction subscription
			if count == 5 && transactionAdded.CompareAndSwap(false, true) {
				fmt.Println("\n📝 Adding transaction subscription after 5 slots...")

				transactionReq := &laserstream.SubscribeRequest{
					Transactions: map[string]*laserstream.SubscribeRequestFilterTransactions{
						"client": {
							Vote:   &[]bool{false}[0],
							Failed: &[]bool{false}[0],
						},
					},
				}

				if err := client.Write(transactionReq); err != nil {
					fmt.Printf("❌ Failed to add transaction subscription: %v\n", err)
				} else {
					fmt.Println("✅ Successfully added transaction subscription")
				}
			}

		case *laserstream.SubscribeUpdate_Transaction:
			fmt.Printf("💸 Transaction update: %s\n", u.Transaction.Transaction.Signature)

			// After receiving some transactions, add a block subscription
			if count >= 15 && blockAdded.CompareAndSwap(false, true) {
				fmt.Println("\n📦 Adding block subscription...")

				blockReq := &laserstream.SubscribeRequest{
					Blocks: map[string]*laserstream.SubscribeRequestFilterBlocks{
						"client": {
							IncludeTransactions: &[]bool{true}[0],
							IncludeAccounts:     &[]bool{false}[0],
							IncludeEntries:      &[]bool{false}[0],
						},
					},
				}

				if err := client.Write(blockReq); err != nil {
					fmt.Printf("❌ Failed to add block subscription: %v\n", err)
				} else {
					fmt.Println("✅ Successfully added block subscription")
				}
			}

		case *laserstream.SubscribeUpdate_Block:
			txCount := 0
			if u.Block.Transactions != nil {
				txCount = len(u.Block.Transactions)
			}
			fmt.Printf("📦 Block update: slot %d, %d transactions\n", u.Block.Slot, txCount)

		default:
			fmt.Printf("📦 Received update: %T\n", u)
		}

		// Stop after 25 messages
		if count >= 25 {
			fmt.Println("\n🛑 Received 25 messages, shutting down...")
			client.Unsubscribe()
			os.Exit(0)
		}
	}

	// Error handler
	errorHandler := func(err error) {
		log.Printf("Stream error: %v", err)
	}

	// Subscribe
	if err := client.Subscribe(initialReq, dataHandler, errorHandler); err != nil {
		log.Fatalf("Failed to subscribe: %v", err)
	}

	fmt.Println("✅ Stream started successfully")

	// Wait for interrupt
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)

	// Add a timeout to demonstrate the feature
	go func() {
		time.Sleep(60 * time.Second)
		fmt.Println("\n⏰ Timeout reached, shutting down...")
		sigChan <- syscall.SIGTERM
	}()

	<-sigChan

	fmt.Println("\n🛑 Shutting down...")
	client.Unsubscribe()
}

File: rust/examples/stream_write_example.rs
use helius_laserstream::{subscribe, LaserstreamConfig, SubscribeRequest, SubscribeRequestFilterSlots};
use futures::StreamExt;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Arc;
use tokio::time::{sleep, Duration};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let endpoint = std::env::var("LASERSTREAM_PRODUCTION_ENDPOINT")
        .unwrap_or_else(|_| "".to_string());
    let api_key = std::env::var("LASERSTREAM_PRODUCTION_API_KEY")
        .expect("LASERSTREAM_PRODUCTION_API_KEY environment variable must be set");

    let config = LaserstreamConfig::new(endpoint, api_key)
        .with_max_reconnect_attempts(5);

    // Initial subscription request - just subscribe to slots
    let mut slots_filter = HashMap::new();
    slots_filter.insert(
        "client".to_string(),
        SubscribeRequestFilterSlots {
            filter_by_commitment: Some(true),
            ..Default::default()
        },
    );

    let initial_request = SubscribeRequest {
        slots: slots_filter,
        commitment: Some(1), // Confirmed
        ..Default::default()
    };

    println!("🚀 Laserstream Bidirectional Stream Example");
    println!("📡 Starting initial subscription to slots...");

    let (stream, handle) = subscribe(config, initial_request);
    let mut stream = Box::pin(stream);

    let message_count = Arc::new(AtomicU32::new(0));
    let handle_clone = handle.clone();
    let count_clone = message_count.clone();

    // Spawn a task to add subscriptions dynamically
    tokio::spawn(async move {
        // Wait for 5 slot updates before adding transaction subscription
        while count_clone.load(Ordering::Relaxed) < 5 {
            sleep(Duration::from_millis(100)).await;
        }

        println!("\n📝 Adding transaction subscription after 5 slots...");
        
        let mut transactions_filter = HashMap::new();
        transactions_filter.insert(
            "client".to_string(),
            helius_laserstream::SubscribeRequestFilterTransactions {
                vote: Some(false),
                failed: Some(false),
                ..Default::default()
            },
        );

        let transaction_request = SubscribeRequest {
            transactions: transactions_filter,
            ..Default::default()
        };

        if let Err(e) = handle_clone.write(transaction_request).await {
            eprintln!("❌ Failed to add transaction subscription: {}", e);
        } else {
            println!("✅ Successfully added transaction subscription");
        }

        // Wait for more messages before adding block subscription
        while count_clone.load(Ordering::Relaxed) < 15 {
            sleep(Duration::from_millis(100)).await;
        }

        println!("\n📦 Adding block subscription...");
        
        let mut blocks_filter = HashMap::new();
        blocks_filter.insert(
            "client".to_string(),
            helius_laserstream::SubscribeRequestFilterBlocks {
                include_transactions: Some(true),
                include_accounts: Some(false),
                include_entries: Some(false),
                ..Default::default()
            },
        );

        let block_request = SubscribeRequest {
            blocks: blocks_filter,
            ..Default::default()
        };

        if let Err(e) = handle_clone.write(block_request).await {
            eprintln!("❌ Failed to add block subscription: {}", e);
        } else {
            println!("✅ Successfully added block subscription");
        }
    });

    // Process the stream
    while let Some(update) = stream.next().await {
        match update {
            Ok(update) => {
                let count = message_count.fetch_add(1, Ordering::Relaxed) + 1;

                match &update.update_oneof {
                    Some(helius_laserstream::grpc::subscribe_update::UpdateOneof::Slot(slot)) => {
                        println!("🎰 Slot update #{}: {}", count, slot.slot);
                    }
                    Some(helius_laserstream::grpc::subscribe_update::UpdateOneof::Transaction(tx)) => {
                        println!("💸 Transaction update: {} bytes", tx.transaction.as_ref().map_or(0, |t| t.data.len()));
                    }
                    Some(helius_laserstream::grpc::subscribe_update::UpdateOneof::Block(block)) => {
                        println!("📦 Block update: slot {}, {} transactions", 
                            block.slot,
                            block.transactions.as_ref().map_or(0, |txs| txs.len())
                        );
                    }
                    _ => {}
                }

                // Stop after 25 messages
                if count >= 25 {
                    println!("\n🛑 Received 25 messages, shutting down...");
                    break;
                }
            }
            Err(e) => {
                eprintln!("❌ Stream error: {}", e);
                // The stream will automatically reconnect
            }
        }
    }

    Ok(())
}